File: 084_Binary Tree Level Order Traversal.cpp
// Binary Tree Level Order Traversal
// Return BFS level order traversal of a binary tree.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

vector<vector<int>> levelOrder(TreeNode* root){
    vector<vector<int>> res;
    if(!root) return res;
    queue<TreeNode*> q; q.push(root);
    while(!q.empty()){
        int n=q.size(); vector<int> row;
        for(int i=0;i<n;++i){ TreeNode* t=q.front(); q.pop(); row.push_back(t->val); if(t->left) q.push(t->left); if(t->right) q.push(t->right);} 
        res.push_back(row);
    }
    return res;
}

int main(){
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto v = levelOrder(root);
    cout<<"Levels:\n";
    for(auto &row:v){ for(int x:row) cout<<x<<" "; cout<<"\n"; }
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: 084_Binary Tree Level Order Traversal.py
"""
Binary Tree Level Order Traversal
Return level order traversal (BFS) of a binary tree's node values.
"""
from typing import Optional, List

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    if not root: return []
    from collections import deque
    q = deque([root]); res=[]
    while q:
        n=len(q); row=[]
        for _ in range(n):
            node=q.popleft(); row.append(node.val)
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        res.append(row)
    return res

if __name__ == '__main__':
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    for row in level_order(root): print(row)

# Time Complexity: O(n)
# Space Complexity: O(n)
